packages/babel-plugin-transform-es2015-destructuring/src/index.js

```diff
@@ -491,16 +491,27 @@ export default function ({ types: t }) {
         }
 
         const nodesOut = [];
-        for (const node of nodes) {
-          const tail = nodesOut[nodesOut.length - 1];
-          if (
-            tail && t.isVariableDeclaration(tail) && t.isVariableDeclaration(node) &&
-            tail.kind === node.kind
-          ) {
-            // Create a single compound let/var rather than many.
-            tail.declarations.push(...node.declarations);
-          } else {
-            nodesOut.push(node);
+
+        // Group all declarations to a single node to avoid
+        // wrong transformation in the `for` initialization.
+        // See https://github.com/babel/babel/issues/4893
+        if (t.isForStatement(parent)) {
```

Comment:
Good idea. Instead of adding new code, we might do this to achieve the same result in a single code path (basically remove the `kind` check and, not so sure, convert `let`/`const` to `var` to avoid the cases when automatically generated vars will be made `let` or `const`):

```js
const nodesOut = [];
for (const node of nodes) {
  const tail = nodesOut[nodesOut.length - 1];
  if (tail && t.isVariableDeclaration(tail) && t.isVariableDeclaration(node)) {
    // Create a single compound declarations
    tail.declarations.push(...node.declarations);
  } else {
    // Question: should we do this??
    if (node.kind === "let" || node.kind === "const") {
      node[t.BLOCK_SCOPED_SYMBOL] = true;
      node.kind = "var";
    }
    // -- end
    nodesOut.push(node);
  }
}
```

My concern when applying this to all cases is this. Let's say we have this declaration:

```js
const { isExternal, children } = this.props;
```

Currently, after destructuring, the code groups into 2 declarations:

```js
var _props = this.props;
const isExternal = _props.isExternal,
      children = _props.children;
```

With the change in this PR, there will only be 1 declaration:

```js
var _props = this.props,
    isExternal = _props.isExternal,
    children = _props.children;
```

This is beyond my knowledge, but is the above safe? Is it the case that var duplication checking and block scoping etc. will be taken of somehow and it doesn't matter that the destructure plugin merges everything into a single `var`?

I have applied the change and run tests. It seems there's no semantic test failures, so appear to be safe. But as I said, this is beyond my knowledge, I don't know if it could break in some case.

Any feedback is appreciated.